executor/join_result_generators.go

```diff
@@ -23,36 +23,66 @@ import (
 )
 
 var (
-	_ joinResultGenerator = &semiJoinResultGenerator{}
-	_ joinResultGenerator = &antiSemiJoinResultGenerator{}
-	_ joinResultGenerator = &leftOuterSemiJoinResultGenerator{}
-	_ joinResultGenerator = &antiLeftOuterSemiJoinResultGenerator{}
-	_ joinResultGenerator = &leftOuterJoinResultGenerator{}
-	_ joinResultGenerator = &rightOuterJoinResultGenerator{}
-	_ joinResultGenerator = &innerJoinResultGenerator{}
+	_ recordJoiner = &semiJoiner{}
+	_ recordJoiner = &antiSemiJoiner{}
+	_ recordJoiner = &leftOuterSemiJoiner{}
+	_ recordJoiner = &antiLeftOuterSemiJoiner{}
+	_ recordJoiner = &leftOuterJoiner{}
+	_ recordJoiner = &rightOuterJoiner{}
+	_ recordJoiner = &innerJoiner{}
 )
 
-// joinResultGenerator is used to generate join results according the join type, see every implementor for detailed information.
-type joinResultGenerator interface {
-	// emit tries to join an outer row with a batch of inner rows.
-	// When inners == nil or inners.Len() == 0, it means that the outer row can not be joined with any inner row:
-	//     1. SemiJoin:              unmatched outer row is ignored.
-	//     2. AntiSemiJoin:          unmatched outer row is appended to the result buffer.
-	//     3. LeftOuterSemiJoin:     unmatched outer row is appended with 0 and appended to the result buffer.
-	//     4. AntiLeftOuterSemiJoin: unmatched outer row is appended with 1 and appended to the result buffer.
-	//     5. LeftOuterJoin:         unmatched outer row is joined with a row of NULLs and appended to the result buffer.
-	//     6. RightOuterJoin:        unmatched outer row is joined with a row of NULLs and appended to the result buffer.
-	//     7. InnerJoin:             unmatched outer row is ignored.
-	// When inners.Len != 0 but all the joined rows are filtered, this means that the outer row is unmatched and the above action is tacked as well.
-	// Otherwise, the outer row is matched and some joined rows is appended to the `chk`.
-	// The size of `chk` is MaxChunkSize at most.
-	emit(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) error
+// recordJoiner is used to generate join results according the join type.
+// A typical instruction flow is:
+//
+//     hasMatch := false
+//     for innerIter.Current() != innerIter.End() {
+//         matched, err := g.tryToMatch(outer, innerIter, chk)
+//         // handle err
+//         hasMatch = hasMatch || matched
+//     }
+//     if !hasMatch {
+//         g.onMissMatch(outer)
+//     }
+//
+// NOTE: This interface is **not** thread-safe.
+type recordJoiner interface {
```

Comment:
It should be `joinRecorder`?

Reply:
`record` means data record, it should be `recordJoiner` or `tupleJoiner`.

Reply:
`record` have both verb and noun, and not used very often in executor package. What about `rowJoiner`.

Reply:
The struct name can be discussed later, we can firstly focus on the code logic.

Reply:
Can we keep the original name `joinResultGenerator` and do the name refactor later?
Because there are too many places to change.